rfcs/text/0003_application_service_mounting.md

```diff
@@ -0,0 +1,259 @@
+      mount(targetDomElement, pluginStartContext, pluginStart) {
+        ReactDOM.render(
+          <MyAppRoot core={pluginStartContext} deps={pluginStart} />,
+          targetDomElement
+        );
+
+        return () => {
+          ReactDOM.unmountComponentAtNode(targetDomElement);
+        };
+      }
+    });
+  }
+}
+```
+
+# Motivation
+
+By having centralized management of applications we can have a true single page
+application. It also gives us a single place to enforce authorization and/or
+licensing constraints on application access.
+
+By making the mounting interface of the ApplicationService generic, we can
+support many different rendering technologies simultaneously to avoid framework
+lock-in.
+
+# Detailed design
+
+This is the bulk of the RFC. Explain the design in enough detail for somebody
+familiar with Kibana to understand, and for somebody familiar with the
+implementation to implement. This should get into specifics and corner-cases,
+and include examples of how the feature is used. Any new terminology should be
+defined here.
+
+## Interface
+
+```ts
+export type Unmount = () => void;
+
+export interface AppSpec {
+  id: string;
+
+  /**
+   * The title of the application.
+   */
+  title: string;
+
+  /**
+   * The root route to mount this application at. Must be unique across all
+   * registered applications.
+   */
+  rootRoute: string;
+
+  /**
+   * A mount function called when the user navigates to this app's `rootRoute`.
+   * @param targetDomElement An HTMLElement to mount the application onto.
+   * @returns An unmounting function that will be called to unmount the application.
+   */
+  mount(targetDomElement: HTMLElement): Unmount
+
+  /**
+   * A EUI iconType that will be used for the app's icon. This icon
+   * takes precendence over the `icon` property.
+   */
+  euiIconType?: string;
+
+  /**
+   * A URL to an image file used as an icon. Used as a fallback
+   * if `euiIconType` is not provided.
+   */
+  icon?: string;
+
+  /**
+   * Custom capabilities defined by the app.
+   */
+  capabilities?: Partial<Capabilities>;
+}
+
+export interface ApplicationSetup {
+  /**
+   * Registers an application with the system.
+   */
+  registerApp(app: AppSpec): void;
+}
+
+export interface ApplicationStart {
+  /**
+   * The UI capabilities for the current user.
+   */
+  capabilities: Capabilties;
+}
+```
+
+## Mounting
+
+When an app is registered via `registerApp`, it must provide a `mount` function
+that will be invoked whenever the window's location has changed from another app
+to this app.
+
+This function is called with an `HTMLELement` for the application
+to render itself to. The application must also return a function that can be
+called by the ApplicationService to unmount the application at the given DOM
+node.
+
+The ApplicationService's `registerApp` method will only be available during the
+*setup* lifecycle event. This allows the system to know when all applications
+have been registered.
```

Comment:
Would be great if application could register and un-register apps at any time. Could you please elaborate on why `registerApp` should be available only inside `.setup` lifecycle?

Reply:
Can you describe the use case you had in mind?

Reply:
@epixa for example, it could be wrapped in React component:

```jsx
<Application id="management" title={t('Management')}>
  <Router>
    <Route match="/index-patterns" comp={SectionIndexPatterns} />
    <Route match="/saved-objects" comp={SectionSavedObjects} />
  </Router>
</Application>
```

Reply:
Application registration is used to send a list of apps to the server for retrieving capabilities. Changing this at runtime would complicate this.

Reply:
@streamich A goal of the new platform is to be framework-agnostic so that any consuming code can use whatever rendering they'd like.

The "routing" provided by the ApplicationService will not be full-featured. Any complex, sub-routing and navigation should be handled by your application's router. You could think of the ApplicationService's routing as just top-level namespacing to handle mounting an application, at which point the application handles the rest.